### PR TITLE
Os 297 system entities

### PR DESCRIPTION
**Pull request by jared-improbable**

### CODE REVIEW

File: Workers/GameLogic/Program.cs [hunk 1]
```diff
@@ -30,9 +28,8 @@ private static async Task<int> Main(string[] args)
 
             IWorkerOptions options = null;
 
-            Parser.Default.ParseArguments<ReceptionistOptions, LocatorOptions>(args)
-                .WithParsed<ReceptionistOptions>(opts => options = opts)
-                .WithParsed<LocatorOptions>(opts => options = opts);
+            Parser.Default.ParseArguments<ReceptionistOptions>(args)
```
Comment (jared-improbable) [inline Workers/GameLogic/Program.cs:31]: Bias to writing managed workers rather than client workers.

File: Workers/GameLogic/Program.cs [hunk 2]
```diff
@@ -105,32 +102,6 @@ private static async Task RunAsync(IWorkerOptions options)
 
         private static void ProcessOpList(OpList opList)
         {
-            foreach (var logOp in opList.OfOpType<LogMessageOp>())
```
Comment (jared-improbable) [inline Workers/GameLogic/Program.cs:108]: They're deprecated anyway, just kill 'em

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs [hunk 3]
```diff
@@ -186,45 +187,63 @@ public void StopSendingMetrics()
             metricsTcs = null;
         }
 
-        private static string GetDevelopmentPlayerIdentityToken(string host, ushort port, string authToken, string playerId, string displayName)
+        private static string GetDevelopmentPlayerIdentityToken(string host, ushort port, bool useInsecureConnection, string authToken, string playerId, string displayName)
         {
-            var pit = DevelopmentAuthentication.CreateDevelopmentPlayerIdentityTokenAsync(
+            using (var pit = DevelopmentAuthentication.CreateDevelopmentPlayerIdentityTokenAsync(
```
Comment (jared-improbable) [inline Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:192]: Memory leak! cc @francissauch

Comment (francissauch) [reply Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs:192]: my bad! good catch :)

File: Workers/GameLogic/Program.cs (hunk 2 above)
Comment (jamiebrynes7) [reply Workers/GameLogic/Program.cs:108]: Wait `LogMessageOp` is deprecated! Brb going to remove this in https://github.com/spatialos/gdk-for-unity too

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs [hunk 4]
```diff
@@ -61,7 +60,7 @@ public static Task<WorkerConnection> ConnectAsync(IWorkerOptions workerOptions,
 
                 case ILocatorOptions locatorOptions:
                     connectionParameters.Network.UseExternalIp = true;
-                    return ConnectAsync(locatorOptions.SpatialOsHost, locatorOptions.SpatialOsPort, connectionParameters, locatorOptions.Token, "Player", "", connectionParameters.WorkerType, taskOptions);
+                    return ConnectAsync(locatorOptions.SpatialOsHost, locatorOptions.SpatialOsPort, locatorOptions.UseInsecureConnection, connectionParameters, locatorOptions.DevToken, locatorOptions.PlayerId, locatorOptions.DisplayName, taskOptions);
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Any reason the `ConnectAsync` overload doesn't take `ILocatorOptions`?

**Review by jamiebrynes7 — APPROVED** [review]
LGTM

File: Improbable/Stdlib/Improbable.Stdlib/Platform/DeploymentWorkerFlags.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        public DeploymentWorkerFlags(string workerType, IEnumerable<Flag> flags)
+        {
+            if (flags == null)
+            {
+                throw new ArgumentNullException(nameof(flags));
+            }
+
+            WorkerType = workerType;
+            Flags = ImmutableArray<Flag>.Empty;
+            Flags = Flags.AddRange(flags);
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/DeploymentWorkerFlags.cs:25]: Can we not initialize this directly? 😱

File: Improbable/Stdlib/Improbable.Stdlib/Platform/LayerConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+using Newtonsoft.Json;
+
+namespace Improbable.Stdlib.Platform
+{
+    public readonly struct LayerConfiguration
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/LayerConfiguration.cs:8]: Worth adding a small note that this is incomplete?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            var tempDir = Path.Combine(rootDir, ".improbable");
+
+            Directory.CreateDirectory(tempDir);
+
+            var splConfigPath = Path.Combine(tempDir, "spl_config.json");
+            Shell.Run(
+                "spatial", "alpha", "config", "convert", "--json_output", "--launch_config", Shell.Escape(startDeployment.LaunchConfigPath),
+                "--main_config", Shell.Escape(startDeployment.ProjectConfigPath), "--output_path", splConfigPath);
+
+            var project = JsonConvert.DeserializeObject<ProjectConfig>(File.ReadAllText(startDeployment.ProjectConfigPath, Encoding.UTF8));
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:63]: Probably worth try-catching to handle breaking changes in the FPL config files?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            // The runtime creates a '.improbable' folder next to the spatialos.json file.
+            var rootDir = Path.GetDirectoryName(startDeployment.ProjectConfigPath);
+            var tempDir = Path.Combine(rootDir, ".improbable");
+
+            Directory.CreateDirectory(tempDir);
+
+            var splConfigPath = Path.Combine(tempDir, "spl_config.json");
+            Shell.Run(
+                "spatial", "alpha", "config", "convert", "--json_output", "--launch_config", Shell.Escape(startDeployment.LaunchConfigPath),
+                "--main_config", Shell.Escape(startDeployment.ProjectConfigPath), "--output_path", splConfigPath);
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:61]: This is not ideal 😕

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            var project = JsonConvert.DeserializeObject<ProjectConfig>(File.ReadAllText(startDeployment.ProjectConfigPath, Encoding.UTF8));
+
+            var needsStart = true;
+            var pidFile = Path.Combine(ToolbeltConfigDir, "spatiald.pid");
+            var lockFile = Path.Combine(ToolbeltConfigDir, "spatiald.lock");
+            if (File.Exists(pidFile) && int.TryParse(File.ReadAllText(pidFile).Trim(), out var pid))
+            {
+                try
+                {
+                    Process.GetProcessById(pid);
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:72]: What if `spatiald` has been started with a different project?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+                Shell.Run("spatial", "service", "start", "--json_output", "--main_config", Shell.Escape(startDeployment.ProjectConfigPath));
+            }
+
+            using (var tcs = new CancellationTokenSource(5000))
+            {
+                var timeout = CallSettings.FromCancellationToken(tcs.Token);
+                var deployments = client.ListDeployments(new ListDeploymentsRequest
+                    { DeploymentName = "local", ProjectName = project.ProjectName }, timeout);
+
+                foreach (var dpl in deployments.Where(dpl => dpl.Status == Deployment.Types.Status.Running))
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: You should be able to add a filter onto the `ListDeploymentsRequest`? Or at least you can when using `spot` (which leads to me believe you can with the platform SDK

Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: What happens if the deployments call hits the time out?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+            var refreshToken = File.ReadAllText(Path.Combine(ToolbeltConfigDir, "oauth2/oauth2_refresh_token"));
+            return new PlatformRefreshTokenCredential(refreshToken);
+        }
+
+        private static PlatformApiChannel GetLocalApiChannel()
+        {
+            return new PlatformApiChannel(GetCredentials(), new PlatformApiEndpoint("localhost", 9876, true));
+        }
+
+        public static Deployment StartLocal(in StartDeploymentOptions startDeployment)
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: Any reason you aren't using async-await here? There should be async functions for the Platform SDK calls

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+                }
+            }).PollUntilCompleted().Result;
+
+            // Apply desired starting worker flags.
+            if (startDeployment.WorkerFlags.Any())
+            {
+                var allWorkerFlags = new RepeatedField<WorkerFlag>();
+
+                var startFlags = startDeployment.WorkerFlags.SelectMany(wf =>
+                    wf.Flags.Select(f => new WorkerFlag { Key = f.Name, Value = f.Value, WorkerType = wf.WorkerType }));
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: Worth providing interop between on the `Platform.DeploymentWorkerFlags` object and this one to tidy this up? 

I.e. - 

```csharp
public IEnumerable<WorkerFlag> ToPlatformSdk()
{
	...
}
```

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+                allWorkerFlags.Add(startFlags);
+
+                var updatedFields = new Deployment{ Name= response.Name, ProjectName = project.ProjectName, Id = response.Id, WorkerFlags = { allWorkerFlags } };
+                client.UpdateDeployment(new UpdateDeploymentRequest { Deployment = updatedFields, UpdateMask = new FieldMask {Paths = { "worker_flags"}}});
+            }
+
+            return response;
+        }
+
+        public static void Stop(in ProjectConfig config)
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]:
```suggestion
        public static void StopAllDeployments(in ProjectConfig config)
```

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+                    }
+                }
+            }
+
+            if (needsStart)
+            {
+                Shell.Run("spatial", "service", "start", "--json_output", "--main_config", Shell.Escape(startDeployment.ProjectConfigPath));
+            }
+
+            using (var tcs = new CancellationTokenSource(5000))
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: Can use the `Stop(...)` function defined below 😄

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+            {
+                var timeout = CallSettings.FromCancellationToken(tcs.Token);
+                var authClient = PlayerAuthServiceClient.Create(credentials: GetCredentials());
+
+                var token = authClient.CreateDevelopmentAuthenticationToken(new CreateDevelopmentAuthenticationTokenRequest
+                {
+                    Description = config.ProjectName,
+                    Lifetime = Duration.FromTimeSpan(TimeSpan.FromMinutes(30)),
+                    ProjectName = config.ProjectName
+                }, timeout);
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: What happens if we hit the timeout?

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+                if (process.ExitCode != 0)
+                {
+                    throw new Exception($"ExitCode {process.ExitCode}: {processStartInfo.FileName} {processStartInfo.Arguments}\n{stdout}\n{stderr}");
+                }
+            }
+
+            public static string Escape(string arg)
+            {
+                return arg.Any(char.IsSeparator) ? $"\"{arg}\"" : arg;
```
Comment (jamiebrynes7) [inline Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:254]: 💯

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs (hunk 4 above)
Comment (jamiebrynes7) [reply Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Bump 😄

File: ci/test.sh [hunk 17]
```diff
@@ -4,6 +4,8 @@ set -euxo pipefail
 
 cd "$(dirname "$0")/../"
 
+ls -lah $IMPROBABLE_CONFIG_DIR/oauth2
```
Comment (jamiebrynes7) [inline ci/test.sh]: Probably remove this now?

**Review by jamiebrynes7 — COMMENTED** [review]
Need a linting pass over the `Platform/Platform.cs` file I think 😄

File: Improbable/Stdlib/Improbable.Stdlib/Platform/DeploymentWorkerFlags.cs (hunk 5 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/DeploymentWorkerFlags.cs:25]: Nope, the API is a bit...yuck.

File: Improbable/Stdlib/Improbable.Stdlib/Platform/LayerConfiguration.cs (hunk 6 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/LayerConfiguration.cs:8]: Done

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 8 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:61]: Nuh uh. FPL needs finishing off.

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 9 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:72]: I'm not worried about it for now.

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 10 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: `TaskCancelledException` is thrown

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 15 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: `TaskCancelledException` is thrown

File: Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs (hunk 4 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs]: Old API - have updated!

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 7 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs:63]: I'd just let exceptions bubble out

File: Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs (hunk 11 above)
Comment (jared-improbable) [reply Improbable/Stdlib/Improbable.Stdlib/Platform/Platform.cs]: Nope, just started as a direct port of the inline versions I was using the now-dead SpotSharp CLI tool

**Pull request merged**